### **FUENTES**:

PetFinder Kaggle:

https://www.kaggle.com/competitions/petfinder-adoption-prediction/data

First Tutorial:

https://towardsdatascience.com/how-to-train-an-image-classifier-in-pytorch-and-use-it-to-perform-basic-inference-on-single-images-99465a1e9bf5

Second Deep Tutorial:

https://rumn.medium.com/part-1-ultimate-guide-to-fine-tuning-in-pytorch-pre-trained-model-and-its-configuration-8990194b71e

Logo Recognition API:

https://heartbeat.comet.ml/logo-recognition-ios-application-using-machine-learning-and-flask-api-aec4eff3be11

Hybrid (multimodal) neural network architecture : Combination of tabular, textual and image inputs:

https://medium.com/@dave.cote.msc/hybrid-multimodal-neural-network-architecture-combination-of-tabular-textual-and-image-inputs-7460a4f82a2e



### **INDICACIONES PREVIAS**:

+ **Git**:
    + Clonamos el repo: root de todos los repos y ponemos git clone "url_repo"
    + Hacemos el checkout de la rama main: git checkout -b new-branch

+ **Poetry**:
    + Instalamos poetry: https://python-poetry.org/docs/
    + Realizamos un Update del pyproject: poetry update
    + Activamos el entorno que creo poetry: poetry shell
    + Intentamos correr una celda, si nos pide seleccionar el environment y no lo vemos en la lista, cerrar y volver abrir VSC

+ **Torch y CUDA**:
    + Verificar que versión pide torch:
        + Versión de torch instalada: poetry show (en mi caso la 1.13.1)
        + Buscar la versión correspondiente en la documentación: https://pytorch.org/get-started/previous-versions/  (en mi caso el 11.7)
    + Instalar CUDA para Torch (buscar la versión correspondiente de CUDA): https://developer.nvidia.com/cuda-11-7-0-download-archive
    + Verificar que CUDA esté funcional: correr en una celda torch.cuda.is_available()

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# SETUP

#  Clonar repo
import os
if not os.path.exists('/content/UA_MDM_Labo2_Grupo10'):
    !git clone https://github.com/sharonpesca-svg/UA_MDM_Labo2_Grupo10.git

# Instalar dependencias ANTES de importar
!pip install -q optuna optuna-dashboard kaleido==0.2.1

# Path al utils
import sys
sys.path.insert(0, '/content/UA_MDM_Labo2_Grupo10/tutoriales')

#  IMPORTS
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, cohen_kappa_score
import shutil, time, copy, datetime
from tqdm import tqdm

import optuna
from optuna.artifacts import FileSystemArtifactStore, upload_artifact

import torch
import torchvision.models as models
import torch.nn as nn
import torch.optim as optim
from torchvision import datasets, transforms
from torch.autograd import Variable
import torch.nn.functional as F
from joblib import load, dump

from utils import plot_confusion_matrix

import sys
sys.path.insert(0, '/content/UA_MDM_Labo2_Grupo10')  # para encontrar augment/

from augment.autoaugment import ImageNetPolicy
from augment.cutout import Cutout

# Verificar GPU
print("CUDA disponible:", torch.cuda.is_available())


Cloning into 'UA_MDM_Labo2_Grupo10'...
remote: Enumerating objects: 160501, done.
remote: Counting objects: 100% (31/31), done.
remote: Compressing objects: 100% (26/26), done.
remote: Total 160501 (delta 12), reused 13 (delta 5), pack-reused 160470 (from 7)
Receiving objects: 100% (160501/160501), 2.05 GiB | 36.22 MiB/s, done.
Resolving deltas: 100% (86130/86130), done.
Updating files: 100% (178605/178605), done.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.9/79.9 MB 10.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 419.5/419.5 kB 38.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.4/8.4 MB 98.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 103.8/103.8 kB 8.7 MB/s eta 0:00:00
CUDA disponible: True


**Seteo el Modelo**

Teoría de Resnet: https://towardsdatascience.com/introduction-to-resnets-c0a830a288a4

In [3]:
# Importo modelo ResNet entrenado en Imagenet
resnet50 = models.resnet50(weights=models.ResNet50_Weights.DEFAULT)
# Modificar la última capa para adaptarse a tu problema específico
num_ftrs = resnet50.fc.in_features
resnet50.fc = torch.nn.Linear(num_ftrs, 5) # Clasificación 5 clases
# Configuro para usar cuda si está disponible
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
resnet50 = resnet50.to(device)
# Instancio del criterio de pérdida CrossEntropyLoss
criterion = nn.CrossEntropyLoss()



Downloading: "https://download.pytorch.org/models/resnet50-11ad3fa6.pth" to /root/.cache/torch/hub/checkpoints/resnet50-11ad3fa6.pth


100%|██████████| 97.8M/97.8M [00:01<00:00, 97.0MB/s]


**Seteo parámetros, directorios y funciones**

In [4]:
# Paths
BASE_DIR = '/content/UA_MDM_Labo2_Grupo10/'

PATH_TO_TRAIN = os.path.join(BASE_DIR, "input/petfinder-adoption-prediction/train/train.csv")
PATH_TO_IMAGES_DIR = os.path.join(BASE_DIR, "input/petfinder-adoption-prediction/train_images")
PATH_TO_TEMP_FILES =  '/content/drive/MyDrive/optuna_temp_artifacts'  #os.path.join(BASE_DIR, "work/optuna_temp_artifacts")
PATH_TO_OPTUNA_ARTIFACTS =  '/content/drive/MyDrive/optuna_artifacts' #os.path.join(BASE_DIR, "work/optuna_artifacts")

MODEL_NAME = '04 ResNet Augment'

MODEL_VERSION = '1.0.0'

# Parametros y variables
CREATE_PYTORCH_DIRECTORIES = 1
SEED = 42
BATCH_SIZE = 32
TEST_SIZE = 0.2
IMAGE_SIZE = 299
CPU_CORES = os.cpu_count()


# Armo el nuevo directorio de train
new_train_directory = os.path.join(BASE_DIR, 'work/train_images_classes')
os.makedirs(new_train_directory, exist_ok=True) # si ya existe el nombre, lo deja como está

# Armo el nuevo directorio de validación
new_val_directory = os.path.join(BASE_DIR, 'work/val_images_classes')
os.makedirs(new_val_directory, exist_ok=True)

# Definir las clases ordenadas
class_names = ['0', '1', '2', '3', '4']

# Mapear las etiquetas de las clases a números enteros consecutivos
class_to_idx = {class_name: i for i, class_name in enumerate(class_names)}

# Creo las carpetas de clases dentro de los directorios
for clase in class_names: # Una para cada clase
   os.makedirs(os.path.join(new_train_directory, str(clase)), exist_ok=True)
   os.makedirs(os.path.join(new_val_directory, str(clase)), exist_ok=True)

# Creo carpetas de Optuna
os.makedirs(PATH_TO_TEMP_FILES, exist_ok=True)
os.makedirs(PATH_TO_OPTUNA_ARTIFACTS, exist_ok=True)


# Funciones para la carga y el preproceso
def resize_to_square(im):
    old_size = im.shape[:2] # old_size is in (height, width) format
    # Calcula el factor de escala necesario para redimensionar la imagen de manera que el lado más largo tenga el tamaño deseado
    ratio = float(IMAGE_SIZE)/max(old_size)
    # Calcula las nuevas dimensiones de la imagen
    new_size = tuple([int(x*ratio) for x in old_size])
    # Redimensiona la imagen con el nuevo tamaño
    im = cv2.resize(im, (new_size[1], new_size[0]))
    # Calcula las diferencias de tamaño y agrega pixeles (color negro) en los extremos para que quede centrada y cuadrada
    delta_w = IMAGE_SIZE - new_size[1]
    delta_h = IMAGE_SIZE - new_size[0]
    top, bottom = delta_h//2, delta_h-(delta_h//2)
    left, right = delta_w//2, delta_w-(delta_w//2)
    color = [0, 0, 0]
    new_image = cv2.copyMakeBorder(im, top, bottom, left, right, cv2.BORDER_CONSTANT,value=color)
    return new_image


def load_image(pet_id):
    path_to_image = os.path.join(PATH_TO_IMAGES_DIR, f'{pet_id}-1.jpg') # Irá a la primera imagen de la mascota
    image = cv2.imread(path_to_image)
    # Convierte la imagen de BGR a RGB porque estos modelos esperan ese orden de canales
    image = cv2.convertScaleAbs(image)
    image= cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    new_image = resize_to_square(image)
    return new_image


In [5]:

def visualize_pet(pet_id):
    path_to_image = os.path.join(PATH_TO_IMAGES_DIR, f'{pet_id}-1.jpg') # Irá a la primera imagen de la mascota
    # Cargar la imagen
    image_to_show = cv2.imread(path_to_image)
    # Convertir a formato RGB
    image_to_show = cv2.cvtColor(image_to_show, cv2.COLOR_BGR2RGB)
    # Visualizar la imagen
    plt.imshow(image_to_show)
    plt.axis('off')  # No mostrar los ejes
    plt.show()

def visualize_image(image):
    # Convierte la imagen a un formato de enteros (CV_8U)
    image = cv2.convertScaleAbs(image)
    image= cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    # Visualizar la imagen
    plt.imshow(image.astype(np.uint8))
    plt.axis('off')  # No mostrar los ejes
    plt.show()


**Cargo y Proceso Data**

Nota: Pytorch necesita que estén las imágenes en los distintos directorios según su clase y su participación en el training

In [6]:
# Cargo
train_df = pd.read_csv(PATH_TO_TRAIN)

# Split para validación
train_data, val_data = train_test_split(train_df,
                               test_size = TEST_SIZE,
                               random_state = SEED,
                               stratify = train_df.AdoptionSpeed)




if CREATE_PYTORCH_DIRECTORIES == 1: # Poner en 0 si ya tengo las carpetas train_images_classes y val_images_classes con las imágenes copiadas
    # Función para copiar las imágenes a los directorios correspondientes
    def copy_imag(data, directorio_destino):
        for index, row in data.iterrows():
            petID = row['PetID']
            adoption_speed = row['AdoptionSpeed']

            # Nombre del archivo de imagen
            nombre_archivo = f"{petID}-1.jpg"

            # Ruta completa de la imagen de origen
            ruta_origen = os.path.join(PATH_TO_IMAGES_DIR, nombre_archivo)

            # Ruta completa del directorio de destino
            ruta_destino = os.path.join(directorio_destino, str(adoption_speed), nombre_archivo)

            # Verificar si el archivo de origen existe
            if os.path.exists(ruta_origen):
                # Copiar el archivo de origen al directorio de destino
                shutil.copy2(ruta_origen, ruta_destino)
        print("Completada la copia a: ",str(directorio_destino))

    # Copiar las imágenes al directorio de train
    copy_imag(train_data, new_train_directory)

    # Copiar las imágenes al directorio de val
    copy_imag(val_data, new_val_directory)

    print("Proceso completado.")

Completada la copia a:  /content/UA_MDM_Labo2_Grupo10/work/train_images_classes
Completada la copia a:  /content/UA_MDM_Labo2_Grupo10/work/val_images_classes
Proceso completado.


In [ ]:
# Genero los DataLoaders - Agregamos las librerias para Augment:
def create_dataloaders(train_directory, val_directory, batch_size, num_workers):
    # Transformaciones de imagen para el conjunto de entrenamiento
    train_transforms = transforms.Compose([
    transforms.Resize((IMAGE_SIZE, IMAGE_SIZE)),
    transforms.RandomHorizontalFlip(),
    ImageNetPolicy(), #Aplica transformaciones aleatorias pero inteligentes (no totalmente al azar)
            #Incluye cosas como: rotaciones, cambios de color/contraste, shear (deformaciones), traslaciones
    transforms.ToTensor(),
    Cutout(n_holes=1, length=16),
    transforms.Normalize(mean=[0.485, 0.456, 0.406],
                         std=[0.229, 0.224, 0.225])
    ])

    # Transformaciones de imagen para el conjunto de validación (sin data augment)
    val_transforms = transforms.Compose([
        transforms.Resize((IMAGE_SIZE, IMAGE_SIZE)),
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.485, 0.456, 0.406],
                             std=[0.229, 0.224, 0.225])
    ])

    # Crear conjuntos de datos para el conjunto de entrenamiento y validación
    conjunto_entrenamiento = datasets.ImageFolder(train_directory, transform=train_transforms)
    conjunto_validacion = datasets.ImageFolder(val_directory, transform=val_transforms)

    # Asignar las clases ordenadas al conjunto de datos
    conjunto_entrenamiento.class_to_idx = {class_name: i for i, class_name in enumerate(class_names)}
    conjunto_validacion.class_to_idx = {class_name: i for i, class_name in enumerate(class_names)}

    # Crear dataloaders para el conjunto de entrenamiento y validación
    train_dataloader = torch.utils.data.DataLoader(conjunto_entrenamiento, batch_size=batch_size, shuffle=True, num_workers=num_workers)
    val_dataloader = torch.utils.data.DataLoader(conjunto_validacion, batch_size=batch_size, shuffle=False, num_workers=num_workers)

    return train_dataloader, val_dataloader

# Aplico las funcion de los DataLoaders
train_dataloader, val_dataloader = create_dataloaders(new_train_directory , new_val_directory , BATCH_SIZE, CPU_CORES)

In [8]:
#Genero una lista de PetIDs con imagen en el orden en que aparecen en el data loader
test_sample_ids = [i[0].split('/')[-1].split('-')[0] for i in val_dataloader.dataset.samples]

In [9]:
def train_val(model, criterion, dataloaders, datasets, device, num_epochs=20, lr=0.001, momentum = 0.9 ,trial=None):

    # Instancio Stochastic Gradient Descent (SGD): Defino el parámetro del Learning Rate (define "el paso" en que avanzan los pesos en cada iteración) y el Momentum (pone innercia a la dirección del gradiente descendiente para que no cambie de dirección en minimos locales)
    optimizer = optim.SGD(resnet50.parameters(), lr=lr, momentum=momentum) # Parámetros default del SGD

    #Inicializo variables
    since = time.time()


    #Inicializo variable de mejor kappa entre trials
    try:
        #Intento obtener el mejor kappa de optuna
        previous_best = study.best_value
    except:
        #Si no hay, seteo -999
        previous_best = -999

    #Inicializo variables de mejor modelo y mejor accuracy y mejor kappa de este trial
    best_model_wts = copy.deepcopy(model.state_dict())
    best_acc = 0.0
    best_kappa =  -999


    for epoch in range(num_epochs):
        print('Epoch {}/{}'.format(epoch, num_epochs - 1))
        print('-' * 10)

        #Inicializo listas de kappa true y predicted y scores para esta epoch
        epoch_kappa_labels_true = []
        epoch_kappa_labels_predicted = []
        epoch_output_scores = []

        #Cada epoch tiene una fase de entrenamiento y validación
        for phase in ['train', 'val']:
            if phase == 'train':
                model.train()  # Set model to training mode
            else:
                model.eval()   # Set model to evaluate mode

            #Inicializo variables de loss y accuracy para esta fase de epoch
            epoch_phase_running_loss = 0.0
            epoch_phase_running_corrects = 0

            # Itero sobre los datos.
            for inputs, labels in tqdm(dataloaders[phase]):
                inputs = inputs.to(device)
                labels = labels.to(device)

                # Zero the parameter gradients
                optimizer.zero_grad()

                # Forward
                # Track history if only in train
                with torch.set_grad_enabled(phase == 'train'):
                    outputs = model(inputs)
                    _, preds = torch.max(outputs, 1)
                    loss = criterion(outputs, labels)

                    # Backward + optimize only if in training phase
                    if phase == 'train':
                        loss.backward()
                        optimizer.step()
                    elif phase == 'val':
                        #Agrego los valores de kappa true y predicted para cada batch en validación
                        epoch_kappa_labels_true.extend(labels.cpu().numpy().tolist())
                        epoch_kappa_labels_predicted.extend(preds.cpu().numpy().tolist())
                        outputs_np = outputs.cpu().numpy()
                        epoch_output_scores.extend([outputs_np[i,:] for i in range(outputs_np.shape[0])])

                # Statistics for each phase
                epoch_phase_running_loss += loss.item() * inputs.size(0)
                epoch_phase_running_corrects += torch.sum(preds == labels.data)

                #END OF BATCH

            epoch_loss = epoch_phase_running_loss / len(datasets[phase])
            epoch_acc = epoch_phase_running_corrects.double() / len(datasets[phase])

            #Calculo el kappa para cada epoch
            if phase == 'train':
                #overall_train_losses.append(epoch_loss)
                current_kappa_score = np.nan
            else:
                #overall_val_losses.append(epoch_loss)
                current_kappa_score = cohen_kappa_score(epoch_kappa_labels_true,
                                  epoch_kappa_labels_predicted,
                                  weights = 'quadratic')



            print(f'{phase.title()} Loss: {epoch_loss:.4f} Acc: {epoch_acc*100:.2f}% Kappa: {current_kappa_score:.3f}')

            # If this is the best Epoch so far -> Deep copy the model
            if phase == 'val' and current_kappa_score > best_kappa:
                best_acc = epoch_acc
                best_kappa = current_kappa_score
                best_model_wts = copy.deepcopy(model.state_dict())


                #Best Epoch within a trial and better than previous trials
                if trial is not None and best_kappa > previous_best:

                    #Save test dataset with predictions
                    predicted_filename = os.path.join(PATH_TO_TEMP_FILES,f'test_{trial.study.study_name}_{trial.number}.joblib')
                    predicted_df = pd.DataFrame({'PetID':test_sample_ids,
                                'pred':epoch_output_scores}).merge(val_data, on='PetID')
                    dump(predicted_df, predicted_filename)

                    #Generate and save CM
                    cm_filename = os.path.join(PATH_TO_TEMP_FILES,f'cm_{trial.study.study_name}_{trial.number}.jpg')
                    plot_confusion_matrix(epoch_kappa_labels_true,epoch_kappa_labels_predicted).write_image(cm_filename)

            #END OF PHASE

        #END OF EPOCH

    time_elapsed = time.time() - since
    print('Training complete in {:.0f}m {:.0f}s'.format(
        time_elapsed // 60, time_elapsed % 60))
    print('Best val Acc: {:.2f}%'.format(best_acc * 100))

    # Load best model weights
    model.load_state_dict(best_model_wts)

    # Save in optuna trial the best test dataset, cm and model weights
    if trial is not None and best_kappa > previous_best:
        upload_artifact(trial, predicted_filename, artifact_store)

        upload_artifact(trial, cm_filename, artifact_store)

        file_name = f'{MODEL_NAME}_{MODEL_VERSION}_{trial.number}.pth'
        model_path = os.path.join(PATH_TO_TEMP_FILES, file_name)
        torch.save(model, model_path) # Podemos guardar solo los pesos si queremos: best_model.state_dict()
        upload_artifact(trial, model_path, artifact_store)

    return model,best_kappa


**Entreno**
Primera prueba: con 7 epoch y batch size 32

batch_size    = 32
epochs        = 7
learning_rate = 0.001
momentum      = 0.9

In [ ]:
best_model,_ = train_val(resnet50, criterion,
                       dataloaders={'train': train_dataloader,
                                    'val': val_dataloader},
                       datasets={'train': train_data, 'val': val_data},
                       device=device,
                       num_epochs=7)
# Guardo el modelo
run_id = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
file_name = f'{MODEL_NAME}_{MODEL_VERSION}_{run_id}.pth'
model_path = os.path.join(PATH_TO_TEMP_FILES, file_name)
torch.save(best_model, model_path) # Podemos guardar solo los pesos si queremos: best_model.state_dict()
print(f'Modelo guardado en {model_path}')

Epoch 0/6
----------


100%|██████████| 367/367 [03:57<00:00,  1.55it/s]


Train Loss: 1.3881 Acc: 33.09% Kappa: nan


100%|██████████| 92/92 [00:20<00:00,  4.39it/s]


Val Loss: 1.3768 Acc: 33.71% Kappa: 0.278
Epoch 1/6
----------


100%|██████████| 367/367 [03:55<00:00,  1.56it/s]


Train Loss: 1.3600 Acc: 35.81% Kappa: nan


100%|██████████| 92/92 [00:22<00:00,  4.14it/s]


Val Loss: 1.3667 Acc: 34.18% Kappa: 0.282
Epoch 2/6
----------


100%|██████████| 367/367 [03:56<00:00,  1.55it/s]


Train Loss: 1.3454 Acc: 36.94% Kappa: nan


100%|██████████| 92/92 [00:21<00:00,  4.28it/s]


Val Loss: 1.3628 Acc: 34.34% Kappa: 0.303
Epoch 3/6
----------


100%|██████████| 367/367 [03:56<00:00,  1.55it/s]


Train Loss: 1.3280 Acc: 38.22% Kappa: nan


100%|██████████| 92/92 [00:21<00:00,  4.36it/s]


Val Loss: 1.3627 Acc: 34.98% Kappa: 0.312
Epoch 4/6
----------


100%|██████████| 367/367 [03:57<00:00,  1.55it/s]


Train Loss: 1.3127 Acc: 39.40% Kappa: nan


100%|██████████| 92/92 [00:20<00:00,  4.40it/s]


Val Loss: 1.3609 Acc: 35.15% Kappa: 0.302
Epoch 5/6
----------


100%|██████████| 367/367 [03:57<00:00,  1.55it/s]


Train Loss: 1.2966 Acc: 40.64% Kappa: nan


100%|██████████| 92/92 [00:21<00:00,  4.20it/s]


Val Loss: 1.3655 Acc: 35.11% Kappa: 0.317
Epoch 6/6
----------


100%|██████████| 367/367 [03:56<00:00,  1.55it/s]


Train Loss: 1.2767 Acc: 41.70% Kappa: nan


100%|██████████| 92/92 [00:21<00:00,  4.31it/s]


Val Loss: 1.3721 Acc: 35.45% Kappa: 0.322
Training complete in 30m 7s
Best val Acc: 35.45%
Modelo guardado en /content/drive/MyDrive/optuna_temp_artifacts/04 ResNet Augment_1.0.0_20260426_090330.pth


Configruacion de Rangos de HP para Optuna:

*   epoch : 5 a 10
*   momentum : 0.7 a 0.99
*   learning rate: 0.00001 a 0.1
*   batch size: 32, 64
*   n_trials = 30



In [10]:
artifact_store = FileSystemArtifactStore(base_path=PATH_TO_OPTUNA_ARTIFACTS)


def optuna_train(trial):

    epochs = trial.suggest_int('epochs', 5, 10)

    lr = trial.suggest_float('lr', 0.00001, 0.1, log=True)

    momentum = trial.suggest_float('momentum',  0.7, 0.99)

    batch_size = trial.suggest_categorical('batch_size', [32, 64])

    # Regenerar dataloaders con el batch_size del trial
    train_dataloader, val_dataloader = create_dataloaders(
        new_train_directory, new_val_directory, batch_size, CPU_CORES
    )

    _,best_score = train_val(resnet50, criterion,
                       dataloaders={'train': train_dataloader,
                                    'val': val_dataloader},
                       datasets={'train': train_data, 'val': val_data},
                       device=device,
                       num_epochs=epochs,
                       lr=lr,
                       momentum = momentum,
                       trial=trial)


    return(best_score)

In [11]:
BASE_DIR = '/content/drive/MyDrive/optuna_artifacts'

study = optuna.create_study(direction='maximize',
                            storage=f"sqlite:///{ os.path.join(BASE_DIR, 'db.sqlite3') }",
                            study_name=f'{MODEL_NAME}_{MODEL_VERSION}',
                            load_if_exists = True)
study.optimize(optuna_train, n_trials=30)

[I 2026-04-27 22:29:27,119] Using an existing study with name '04 ResNet Augment_1.0.0' instead of creating a new one.


Epoch 0/4
----------


100%|██████████| 184/184 [03:54<00:00,  1.28s/it]


Train Loss: 1.4046 Acc: 31.90% Kappa: nan


100%|██████████| 46/46 [00:20<00:00,  2.19it/s]


Val Loss: 1.3735 Acc: 33.94% Kappa: 0.284
Epoch 1/4
----------


100%|██████████| 184/184 [04:07<00:00,  1.34s/it]


Train Loss: 1.3508 Acc: 36.00% Kappa: nan


100%|██████████| 46/46 [00:21<00:00,  2.10it/s]


Val Loss: 1.3575 Acc: 35.85% Kappa: 0.317
Epoch 2/4
----------


100%|██████████| 184/184 [04:07<00:00,  1.35s/it]


Train Loss: 1.3186 Acc: 38.61% Kappa: nan


100%|██████████| 46/46 [00:21<00:00,  2.17it/s]


Val Loss: 1.3643 Acc: 35.48% Kappa: 0.314
Epoch 3/4
----------


100%|██████████| 184/184 [04:07<00:00,  1.35s/it]


Train Loss: 1.2876 Acc: 41.38% Kappa: nan


100%|██████████| 46/46 [00:21<00:00,  2.17it/s]


Val Loss: 1.3751 Acc: 35.71% Kappa: 0.308
Epoch 4/4
----------


100%|██████████| 184/184 [04:07<00:00,  1.35s/it]


Train Loss: 1.2541 Acc: 42.97% Kappa: nan


100%|██████████| 46/46 [00:21<00:00,  2.15it/s]
[I 2026-04-27 22:51:39,695] Trial 31 finished with value: 0.3171140450168324 and parameters: {'epochs': 5, 'lr': 0.003575790816439069, 'momentum': 0.9114967403334514, 'batch_size': 64}. Best is trial 1 with value: 0.3291620227814408.


Val Loss: 1.3789 Acc: 35.51% Kappa: 0.310
Training complete in 22m 12s
Best val Acc: 35.85%
Epoch 0/5
----------


100%|██████████| 184/184 [04:07<00:00,  1.35s/it]


Train Loss: 1.3426 Acc: 37.19% Kappa: nan


100%|██████████| 46/46 [00:22<00:00,  2.08it/s]


Val Loss: 1.3733 Acc: 34.44% Kappa: 0.273
Epoch 1/5
----------


100%|██████████| 184/184 [04:07<00:00,  1.34s/it]


Train Loss: 1.2962 Acc: 40.21% Kappa: nan


100%|██████████| 46/46 [00:21<00:00,  2.14it/s]


Val Loss: 1.4040 Acc: 34.41% Kappa: 0.300
Epoch 2/5
----------


 35%|███▍      | 64/184 [01:28<02:46,  1.39s/it]
[W 2026-04-27 23:02:07,241] Trial 32 failed with parameters: {'epochs': 6, 'lr': 0.01639174743423472, 'momentum': 0.8698339374932522, 'batch_size': 64} because of the following error: KeyboardInterrupt().
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/optuna/study/_optimize.py", line 206, in _run_trial
    value_or_values = func(trial)
                      ^^^^^^^^^^^
  File "/tmp/ipykernel_10571/4196110891.py", line 19, in optuna_train
    _,best_score = train_val(resnet50, criterion,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/ipykernel_10571/1232547427.py", line 71, in train_val
    epoch_phase_running_loss += loss.item() * inputs.size(0)
                                ^^^^^^^^^^^
KeyboardInterrupt
[W 2026-04-27 23:02:07,243] Trial 32 failed with value None.


KeyboardInterrupt: 

In [1]:
import os, subprocess, threading, time, re
from optuna_dashboard import run_server

# Liberar puerto 8080 en Windows
os.system("for /f \"tokens=5\" %a in ('netstat -ano ^| findstr :8080') do taskkill /PID %a /F 2>nul")
time.sleep(1)

db_path = "sqlite:///" + os.path.abspath("../work/optuna_artifacts/db.sqlite3").replace("\\", "/")

thread = threading.Thread(
    target=run_server,
    args=(db_path,),
    kwargs={"port": 8080},
    daemon=True
)
thread.start()
time.sleep(2)

if not os.path.exists("cloudflared.exe"):
    os.system("curl -L -o cloudflared.exe https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-windows-amd64.exe")

tunnel = subprocess.Popen(
    ["cloudflared.exe", "tunnel", "--url", "http://localhost:8080"],
    stdout=subprocess.PIPE,
    stderr=subprocess.PIPE
)

for line in tunnel.stderr:
    line = line.decode()
    match = re.search(r'https://[a-z0-9\-]+\.trycloudflare\.com', line)
    if match:
        print("Dashboard en:", match.group(0))
        break


c:\Users\u581537.TELECOM.000\.conda\envs\ldi2\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
Bottle v0.13.4 server starting up (using WSGIRefServer())...
Listening on http://localhost:8080/
Hit Ctrl-C to quit.



Dashboard en: https://continent-disco-laboratories-census.trycloudflare.com


127.0.0.1 - - [29/Apr/2026 20:44:15] "GET / HTTP/1.1" 302 0
127.0.0.1 - - [29/Apr/2026 20:44:15] "GET /dashboard HTTP/1.1" 200 4145
127.0.0.1 - - [29/Apr/2026 20:44:15] "GET /static/bundle.js HTTP/1.1" 200 4159096
127.0.0.1 - - [29/Apr/2026 20:44:16] "GET /api/meta HTTP/1.1" 200 113
127.0.0.1 - - [29/Apr/2026 20:44:16] "GET /api/studies HTTP/1.1" 200 273
c:\Users\u581537.TELECOM.000\.conda\envs\ldi2\Lib\site-packages\optuna_dashboard\_importance.py:71: ExperimentalWarning: PedAnovaImportanceEvaluator is experimental (supported from v3.6.0). The interface can change in the future.
  study, target=lambda t: t.values[objective_id], evaluator=PedAnovaImportanceEvaluator()
c:\Users\u581537.TELECOM.000\.conda\envs\ldi2\Lib\site-packages\optuna\importance\_ped_anova\evaluator.py:183: UserWarning: PedAnovaImportanceEvaluator computes the importances of params to achieve low `target` values. If this is not what you want, please modify target, e.g., by multiplying the output by -1.
  optuna_warn

Análisis del dashboard de Optuna:

El mejor trial fue el trial 1 con kappa ~0.329. Dicho valor nunca fue superado. 

El grafico de Importancia de HP en el dashboard de Optuna muestra que epochs (0.56) es el HP más importante por lejos. Cuántas vueltas entrena el modelo impacto más que cualquier otro parámetro.
El learning rate (0.40) es el segundo más importante, la tasa de aprendizaje también pesa mucho.
momentum (0.05) y batch_size (0.01) prácticamente no aportan para los valores configurados (32 y 64).

Conclusiones:

Epochs mas altos (8-10) dan mejores valores de kappa. 

En learning rate los mejores kappa se obtienen con valores intermedios (~0.001-0.01). Los extremos muy bajos o muy altos dan kappa malo.

Momentum no muestra patrón claro, confirma que no importa.

Batch size (32, 64) prácticamente no tiene efecto en la performance de la predicción. 
